File: packages/gbnf-experiment/data/20260909T015142Z_f8ca89ff/ported_implementation/grammar_graph/generic_set.py
from __future__ import annotations

from typing import Callable, Dict, Generic, Iterator, Optional, TypeVar

__all__ = ["GenericSet"]

T = TypeVar("T")
K = TypeVar("K")


class GenericSet(Generic[T, K]):
    """A set that dedupes on a derived key while preserving object identity.

    Mirrors the reference implementation, which pairs a `Map<K, T>` with a
    `Set<T>`; membership on the latter is by reference, which is reproduced here
    with an identity-keyed dict.
    """

    def __init__(self, get_key: Callable[[T], K]):
        self.get_key = get_key
        self._keys: Dict[K, T] = {}
        self._set: Dict[int, T] = {}

    def add(self, el: T) -> None:
        key = self.get_key(el)
        if key not in self._keys:
            self._keys[key] = el
            self._set[id(el)] = el

    def delete(self, el: T) -> None:
        key = self.get_key(el)
        ref = self._keys.get(key)
        if ref is None:
            raise Exception("Could not get ref")
        del self._keys[key]
        self._set.pop(id(ref), None)

    def has(self, el: T) -> bool:
        return id(el) in self._set

    def get(self, el: T) -> Optional[T]:
        return self._keys.get(self.get_key(el))

    def __iter__(self) -> Iterator[T]:
        return iter(list(self._set.values()))

    def __contains__(self, el: object) -> bool:
        return id(el) in self._set

    @property
    def size(self) -> int:
        return len(self._set)

    def __len__(self) -> int:
        return len(self._set)

    def __repr__(self) -> str:
        return f"GenericSet(size={self.size})"

File: proposed-projects/round-trip-experiment/reverse/20260909T015142Z_f8ca89ff/20260910T053604Z_b86749e4/ported_implementation/src/grammar-graph/generic-set.ts
/**
 * A set that dedupes on a derived key while preserving object identity.
 *
 * The `Map` holds the canonical element for each key; the `Set` answers
 * membership by reference.
 */
export class GenericSet<T, K> {
  private keys = new Map<K, T>();
  private set = new Set<T>();

  constructor(public getKey: (el: T) => K) {}

  public add(el: T): void {
    const key = this.getKey(el);
    if (!this.keys.has(key)) {
      this.keys.set(key, el);
      this.set.add(el);
    }
  }

  public delete(el: T): void {
    const key = this.getKey(el);
    const ref = this.keys.get(key);
    if (ref === undefined) {
      throw new Error('Could not get ref');
    }
    this.keys.delete(key);
    this.set.delete(ref);
  }

  public has(el: T): boolean {
    return this.set.has(el);
  }

  public get(el: T): T | undefined {
    return this.keys.get(this.getKey(el));
  }

  public [Symbol.iterator](): IterableIterator<T> {
    return this.set.values();
  }

  public get size(): number {
    return this.set.size;
  }
}
